multiset_test.go

```diff
@@ -0,0 +1,315 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+		m.Add(data)
+	}
+	if m.Finalize() != zeroHash {
+		t.Fatalf("m was expected to be zero hash, but was %s instead", m.Finalize())
+	}
+
+	// Here we first remove an element from an empty multiset, and then add some other
+	// elements, and then we create a new empty multiset, then we add the same elements
+	// we added to the previous multiset, and then we remove the same element we remove
+	// the same element we removed from the previous multiset. According to commutativity
+	// laws, the result should be the same.
+	removeIndex := 0
+	removeData, err := hex.DecodeString(testVectors[removeIndex].dataElementHex)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	m1 := NewMultiset()
+	m1.Remove(removeData)
+
+	for i, test := range testVectors {
+		if i != removeIndex {
+			data, err := hex.DecodeString(test.dataElementHex)
+			if err != nil {
+				t.Fatal(err)
+			}
+			m1.Add(data)
+		}
+	}
+
+	m2 := NewMultiset()
+	for i, test := range testVectors {
+		if i != removeIndex {
+			data, err := hex.DecodeString(test.dataElementHex)
+			if err != nil {
+				t.Fatal(err)
+			}
+			m2.Add(data)
+		}
+	}
+	m2.Remove(removeData)
+
+	if m1.Finalize() != m2.Finalize() {
+		t.Fatalf("m1 and m2 was exepcted to have the same hash, but got instead m1 %s and m2 %s", m1.Finalize(), m2.Finalize())
+	}
+}
+
+func TestParseMultiSetFail(t *testing.T) {
+	data := [64]byte{}
+	copy(data[:], Secp256k1Order.Bytes())
+	_, err := ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("Shouldn't be able to parse a multiset bigger with x bigger than the field size")
+	}
+	data = [64]byte{}
+	copy(data[32:], Secp256k1Order.Bytes())
+	_, err = ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("Shouldn't be able to parse a multiset bigger with y bigger than the field size")
+	}
+	set := NewMultiset()
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("Failed generating random data %v %d", err, n)
+	}
+	set.Add(data[:])
+
+}
+
+func TestMultiSet_Reset(t *testing.T) {
+	set := NewMultiset()
+	emptySet := NewMultiset()
+	data := [100]byte{}
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("Failed generating random data %v %d", err, n)
+	}
+	set.Add(data[:])
+	if set.Finalize() == emptySet.Finalize() {
+		t.Errorf("Set shouldn't be empty")
+	}
+	set.Reset()
+	if set.Finalize() != emptySet.Finalize() {
+		t.Errorf("Set should be empty")
+	}
+}
+
+func TestMultiSetAddRemove(t *testing.T) {
+	list := [LoopsN][100]byte{}
+	set := NewMultiset()
+	set2 := set
+	serializedEmpty := set.Serialize()
+	for i := 0; i < LoopsN; i++ {
+		data := [100]byte{}
+		n, err := rand.Read(data[:])
+		if err != nil || n != len(data) {
+			t.Fatalf("Failed generating random data %v %d", err, n)
+		}
+		set.Add(data[:])
+		list[i] = data
+	}
+	if set.Finalize() == set2.Finalize() {
+		t.Errorf("sets are the same when they should be different: set %x\n", set.Finalize())
```

Comment:
Please print both sets.

Reply:
The `if` only executes if they're the same though

Reply:
Oops, yeah, my bad.